File: dungeons-1.21.11/src/main/java/fishmod/utils/data/ScoreboardUtil.java
package fishmod.utils.data;

import net.minecraft.client.MinecraftClient;
import net.minecraft.client.network.PlayerListEntry;

public class ScoreboardUtil {

    public static String getCurrentClass() {
        ClassInfo info = getClassInfo();
        return info != null ? info.className : null;
    }

    public static int getCurrentClassLevel() {
        ClassInfo info = getClassInfo();
        return info != null ? info.level : 0;
    }

    private static ClassInfo getClassInfo() {
        MinecraftClient mc = MinecraftClient.getInstance();
        if (mc == null || mc.player == null || mc.getNetworkHandler() == null)
            return null;

        String myName = mc.player.getName().getString();

        for (PlayerListEntry entry : mc.getNetworkHandler().getPlayerList()) {
            String profileName = entry.getProfile().name();
            if (!profileName.equals(myName)) continue;

            String display = entry.getDisplayName() != null
                    ? entry.getDisplayName().getString()
                    : profileName;

            // Look for "(Mage XLIX)"
            int start = display.indexOf("(");
            int end = display.indexOf(")");
            if (start == -1 || end == -1) return null;

            String inside = display.substring(start + 1, end).trim();
            // inside = "Mage XLIX"

            String[] parts = inside.split(" ");
            if (parts.length != 2) return null;

            String className = parts[0];
            String roman = parts[1];

            int level = romanToInt(roman);

            return new ClassInfo(className, level);
        }

        return null;
    }

    private static int romanToInt(String roman) {
        int sum = 0;
        int prev = 0;

        for (int i = roman.length() - 1; i >= 0; i--) {
            int val = romanValue(roman.charAt(i));
            if (val < prev) sum -= val;
            else sum += val;
            prev = val;
        }

        return sum;
    }

    private static int romanValue(char c) {
        return switch (c) {
            case 'I' -> 1;
            case 'V' -> 5;
            case 'X' -> 10;
            case 'L' -> 50;
            case 'C' -> 100;
            case 'D' -> 500;
            case 'M' -> 1000;
            default -> 0;
        };
    }

    private record ClassInfo(String className, int level) {}
}

File: dungeons-1.21.11/src/main/kotlin/fishmod/utils/data/ScoreboardUtil.kt
package fishmod.utils.data

import net.minecraft.client.MinecraftClient

object ScoreboardUtil {

    private class ClassInfo(val className: String, val level: Int)

    @JvmStatic
    fun getCurrentClass(): String? {
        val info = getClassInfo()
        return info?.className
    }

    @JvmStatic
    fun getCurrentClassLevel(): Int {
        val info = getClassInfo()
        return info?.level ?: 0
    }

    private fun getClassInfo(): ClassInfo? {
        val mc = MinecraftClient.getInstance()
        if (mc == null || mc.player == null || mc.networkHandler == null) return null

        val myName = mc.player!!.name.string

        for (entry in mc.networkHandler!!.playerList) {
            val profileName = entry.profile.name
            if (profileName != myName) continue

            val display = entry.displayName?.string ?: profileName

            // Look for "(Mage XLIX)"
            val start = display.indexOf("(")
            val end = display.indexOf(")")
            if (start == -1 || end == -1) return null

            val inside = display.substring(start + 1, end).trim()
            // inside = "Mage XLIX"

            val parts = inside.split(" ")
            if (parts.size != 2) return null

            val className = parts[0]
            val roman = parts[1]

            val level = romanToInt(roman)

            return ClassInfo(className, level)
        }

        return null
    }

    private fun romanToInt(roman: String): Int {
        var sum = 0
        var prev = 0

        for (i in roman.length - 1 downTo 0) {
            val value = romanValue(roman[i])
            if (value < prev) sum -= value else sum += value
            prev = value
        }

        return sum
    }

    private fun romanValue(c: Char): Int {
        return when (c) {
            'I' -> 1
            'V' -> 5
            'X' -> 10
            'L' -> 50
            'C' -> 100
            'D' -> 500
            'M' -> 1000
            else -> 0
        }
    }
}
